### ISSUE TITLE
RequestBody + Servant results in EtaException

### ISSUE THREAD

**Issue by NickSeagull** [typelead/eta#626]
I have some custom monad defined for a servant application I'm working on, the problem is that when I try to make a request to an endpoint with a request body it result in this [ugly stacktrace](http://lpaste.net/361990).

My custom monad is defined like

```haskell
data AppConfig = AppConfig
  { clientId :: Maybe String
  , clientSecret :: Maybe String
  }

newtype AppT monad return = AppT
  { runApp :: ReaderT AppConfig (ExceptT ServantErr monad) return
  } deriving ( Functor
             , Applicative
             , Monad
             , MonadReader AppConfig
             , MonadError ServantErr
             , MonadIO
             )

type App = AppT IO
```

And the helper functions to make it work with Servant:

```haskell
type AppAPI = Web.API :<|> Raw


app :: AppConfig -> Application
app cfg = serve (Proxy :: Proxy AppAPI) (appToServer cfg)


appToServer :: AppConfig -> Server AppAPI
appToServer cfg = enter (convertApp cfg >>> NT Handler) handler :<|> serveDirectoryFileServer "static"


convertApp :: AppConfig -> AppT monad :~> ExceptT ServantErr monad
convertApp cfg = runReaderTNat cfg <<< NT runApp
```

The endpoint is defined like

```haskell
data ClientEvent
  = AppStarted
  | LoginButtonClicked
  deriving (Generic)

instance FromJSON ClientEvent
instance ToJSON ClientEvent


data ServerEvent
  = LoadedData ClientId
  | RequestSuccessful String
  deriving (Generic)

instance FromJSON ServerEvent
instance ToJSON ServerEvent

newtype ClientId = ClientId (Maybe String)
  deriving (Generic)

instance FromJSON ClientId
instance ToJSON ClientId

type API =
  "event"
  :> ReqBody '[JSON] ClientEvent
  :> Post '[JSON] ServerEvent

handler :: ClientEvent -> App ServerEvent
handler AppStarted = do
  Log.debug "AppStarted"
  cid <- asks clientId
  return (LoadedData $ ClientId cid)
```

**Comment by NickSeagull** [comment]
I've uploaded the code that reproduces the bug to [this repository](https://github.com/NickSeagull/servant-eta-transformer)

**Comment by NickSeagull** [comment]
I've just updated the code to use the regular Servant `Handler` type, it produces the same stack trace.

**Comment by jneira** [comment]
Hi, i have detected a problem in wai-servlet that maybe it is related with the issue, cause it is about the request body. I think the error can be caused when aeson tries to consume it.
I filled [an issue](https://github.com/jneira/wai-servlet/issues/12) in wai servlet about the problem, that maybe it is related with eta foreign pointers

**Comment by rahulmutt** [comment]
This is fixed in `wai-servlet-0.1.4.0`. There are a subtle change in the new MemoryManager in v0.7 which breaks one of the assumptions inside of `wai-servlet`.

Run the following to fix it:
```
etlas update
etlas install wai-servlet-0.1.4.0
etlas install wai-servlet-handler-jetty --reinstall
```

**Comment by NickSeagull** [comment]
Thanks!
